lib/inbox.rb

```diff
@@ -22,59 +22,69 @@
 
 module Inbox
   Error = Class.new(::StandardError)
-  AccessDenied = Class.new(Error)
-  ResourceNotFound = Class.new(Error)
   NoAuthToken = Class.new(Error)
   UnexpectedAccountAction = Class.new(Error)
   UnexpectedResponse = Class.new(Error)
   class APIError < Error
-    attr_accessor :error_type
+    attr_accessor :type
+    attr_accessor :message
     attr_accessor :server_error
 
-    def initialize(type, error, server_error = nil)
-      super(error)
-      self.error_type = type
+    def initialize(type, message, server_error = nil)
+      super(message)
+      self.type = type
+      self.message = message 
       self.server_error = server_error
     end
   end
+  AccessDenied = Class.new(APIError)
+  ResourceNotFound = Class.new(APIError)
   InvalidRequest = Class.new(APIError)
   MessageRejected = Class.new(APIError)
   SendingQuotaExceeded = Class.new(APIError)
   ServiceUnavailable = Class.new(APIError)
-
-  def self.interpret_http_status(result)
-    # Handle HTTP errors and RestClient errors
-    raise ResourceNotFound.new if result.code.to_i == 404
-    raise AccessDenied.new if result.code.to_i == 403
-  end
+  BadGateway = Class.new(APIError)
+  InternalError = Class.new(APIError)
+  MailProviderError = Class.new(APIError)
 
   HTTP_CODE_TO_EXCEPTIONS = {
     400 => InvalidRequest,
     402 => MessageRejected,
     403 => AccessDenied,
     404 => ResourceNotFound,
+    422 => MailProviderError,
     429 => SendingQuotaExceeded,
-    503 => ServiceUnavailable
+    500 => InternalError,
+    502 => BadGateway,
+    503 => ServiceUnavailable,
   }.freeze
 
   def self.http_code_to_exception(http_code)
     HTTP_CODE_TO_EXCEPTIONS.fetch(http_code, APIError)
   end
 
   def self.interpret_response(result, result_content, options = {})
-    # Handle HTTP errors
-    self.interpret_http_status(result)
 
-    # Handle content expectation errors
+    # We expected a certain kind of object, but the API didn't return anything
     raise UnexpectedResponse.new if options[:expected_class] && result_content.empty?
-    json = options[:result_parsed]? result_content : JSON.parse(result_content)
-    if json.is_a?(Hash) && (json['type'] == 'api_error' or json['type'] == 'invalid_request_error')
+
+    # If it's already parsed, or if we've received an actual raw payload on success, don't parse
+    if options[:result_parsed] || (options[:raw_response] && result.code.to_i == 200)
+      response = result_content
```

Comment:
I think you can just `return result_content` here. No need to have this Pascal-style single-exit thing.